### ISSUE TITLE
top_n() not working on db backends

### ISSUE THREAD

**Issue by javierluraschi** [tidyverse/dplyr#4467]
```r
dbplyr::memdb_frame(a = 1:3, b = 2:4) %>%
  count(a) %>%
  top_n(10)
```
```
Selecting by n
Error: Base operators are not defined for quosures.
Do you need to unquote the quosure?

  # Bad:
  myquosure > rhs

  # Good:
  !!myquosure > rhs
Call `rlang::last_error()` to see a backtrace
```
```
21: stop(cnd)
20: abort(paste_line("Base operators are not defined for quosures.", 
        "Do you need to unquote the quosure?", "", "  # Bad:", bad, 
        "", "  # Good:", good, ))
19: Ops.quosure(.n, 0)
18: eval_bare(call[[2]], env)
17: partial_eval_call(call, vars, env)
16: partial_eval(get_expr(x), vars = vars, env = get_env(x))
15: new_quosure(partial_eval(get_expr(x), vars = vars, env = get_env(x)), 
        get_env(x))
14: FUN(X[[i]], ...)
13: lapply(dots, function(x) {
        new_quosure(partial_eval(get_expr(x), vars = vars, env = get_env(x)), 
            get_env(x))
    })
12: partial_eval_dots(dots, vars = op_vars(.data))
11: filter.tbl_lazy(x, !!pred)
10: filter(x, !!pred)
9: top_n(., 10)
8: function_list[[k]](value)
7: withVisible(function_list[[k]](value))
6: freduce(value, `_function_list`)
5: `_fseq`(`_lhs`)
4: eval(quote(`_fseq`(`_lhs`)), env, env)
3: eval(quote(`_fseq`(`_lhs`)), env, env)
2: withVisible(eval(quote(`_fseq`(`_lhs`)), env, env))
1: dbplyr::memdb_frame(a = 1:3, b = 2:4) %>% count(a) %>% top_n(10)
```

Thought his might have been related to `rlang` but not really, repros with `rlang 0.3.4` and `rlang 0.4.0`

**Comment by romainfrancois** [comment]
@lionel- I go a bit further if I revert `top_n()` to this: 

```r
top_n <- function(x, n, wt) {
  wt <- enquo(wt)
  if (quo_is_missing(wt)) {
    vars <- tbl_vars(x)
    wt_name <- vars[length(vars)]
    inform(glue("Selecting by ", wt_name))
    wt <- sym(wt_name)
  }

  filter(x, !!expr({
    .n <- {{ n }}
    if (.n > 0) {
      min_rank(desc(!!wt)) <= .n
    } else {
      min_rank(!!wt) <= abs(.n)
    }
  }))
}
```

i.e. without using an extra function that apparently was not in scope for the way `dbplyr` works. 

that still gives me this though :

``` r
library(dplyr, warn.conflicts = FALSE)

dbplyr::memdb_frame(a = 1:3, b = 2:4) %>% count(a)  %>% top_n(2)  
#> Selecting by n
#> Warning: `lang_name()` is deprecated as of rlang 0.2.0.
#> Please use `call_name()` instead.
#> This warning is displayed once per session.
#> Warning: `lang()` is deprecated as of rlang 0.2.0.
#> Please use `call2()` instead.
#> This warning is displayed once per session.
#> Error: Window function `RANK()` is not supported by this database
```

because: 

``` r
library(dplyr, warn.conflicts = FALSE)

dbplyr::memdb_frame(a = 1:3, b = 2:4) %>% count(a)  %>% filter(min_rank(desc(n)) <= 2)
#> Warning: `lang_name()` is deprecated as of rlang 0.2.0.
#> Please use `call_name()` instead.
#> This warning is displayed once per session.
#> Warning: `lang()` is deprecated as of rlang 0.2.0.
#> Please use `call2()` instead.
#> This warning is displayed once per session.
#> Error: Window function `RANK()` is not supported by this database
```

**Comment by lionel-** [comment]
Right, translation to other backends is a good reason to program with expressions rather than functions.

Are you sure you need the `!!expr()`?

```r
  filter(x, {
    .n <- {{ n }}
    if (.n > 0) {
      min_rank(desc(!!wt)) <= .n
    } else {
      min_rank(!!wt) <= abs(.n)
    }
  })
```

I also would use a non-syntactic symbols to avoid overriding data mask symbols, like  `_n`.

**Comment by jmerone** [comment]
@romainfrancois @javierluraschi  Is this issues still being addressed? I see the PR for the change hasn't been updated in over a month?

**Comment by krlmlr** [comment]
Why not use/forward to `head()` ?

**Comment by jmerone** [comment]
@krlmlr I have used `head()` as a work around. However, there is a lot of legacy code that used `top_n`, I would like to avoid having to completely rewrite.

**Comment by dgrtwo** [comment]
`top_n()` doesn't work in Redshift either, and can't be replaced with `head()` because the latter doesn't respect groups.

FYI, error for Redshift is:

```
Selecting by clientid
Error in postgresqlExecStatement(conn, statement, ...) : 
  RS-DBI driver: (could not Retrieve the result : ERROR:  function top_n_rank(integer, character varying) does not exist
HINT:  No function matches the given name and argument types. You may need to add explicit type casts.
)
```

**Comment by fh-jgutman** [comment]
> `top_n()` doesn't work in Redshift either, and can't be replaced with `head()` because the latter doesn't respect groups.

Good point! In the meantime, 
`filter(min_rank(var) <= n)` does work in Postgres/Redshift and should respect groups

**Comment by hadley** [comment]
I think we may need to revert #4417, or make `top_n()` generic so that we can override in dbplyr.

**Comment by hadley** [comment]
Or possibly move the `if` outside of the expression when `n` is a constant.

**Comment by hadley** [comment]
This will be resolved by #4687 + https://github.com/tidyverse/dbplyr/issues/394 through the introduction of new `slice_min()` and `slice_max()` functions, which also allow us to resolve some interface issues with `top_n()`.
